### PR TITLE
feat(runtime): automatically insert `key` attrs during compilation

### LINKED ISSUE #1968: Parent with Child Slot rerender bug
**Stencil version:**

```
 @stencil/core@1.7.4
 ---
 update: 15.11.2019
 ---
 @stencil/core@1.8.0
---
 update: 15.01.2020
 ---
 @stencil/core@1.8.5
```

**I'm submitting a:**

[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://stencil-worldwide.herokuapp.com/ or https://forum.ionicframework.com/

**Current behavior:**
Here is a demo: https://dutscher.github.io/stencilslots/

Here is the code which doesnt work from <my-parent /> with comments
```tsx
render() {
    const classes = [];

    if (this.disabled) {
      classes.push(`${this.componentName}--disabled`);
    }

    return (
      <Host class={classes.join(' ')}>
        Demonstrate slot rerender bug
        {/* this dont rerender slot */}
        <div>
          {/* remove condition: works */}
          {/* wrap with <div />: works */}
          {this.disabled && !!this.tag && (
            <div class="tag" innerHTML={this.tag}/>
          )}
          {/* remove div: works */}
          <div>
            before slot&gt;
            <slot/>
            &lt;after slot
          </div>
        </div>
      </Host>
    );
  }
```
The `<slot />` renders first very well, when a rerender in the parent happens, the `<slot />` going away. There is a ugly workarround, you can put a `<div />`. But logical you dont need this `<div/>`.
https://github.com/dutscher/stencilslots/blob/master/src/components/parent-workaround/parent.tsx

I got a scenario there i need more than one `<div />`.

I guess its a vdom diff problem.

**Expected behavior:**
TSX works without extra empty `<div />` and the `<slot />` is rendered after parent is rerendered.

**Steps to reproduce:**
https://dutscher.github.io/stencilslots/

**Related code:**
https://github.com/dutscher/stencilslots/blob/master/src/components/parent-broken/parent.tsx
https://github.com/dutscher/stencilslots/blob/master/src/components/parent-workaround/parent.tsx

Cheers and thanks for the attention :)

### LINKED ISSUE #5263: bug: Conditional slot rendering removes always the last slot element, using key is a workaround
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

We were using 4.8.2 but it also happens with 4.10.0

### Current Behavior

It occurs that always the last slot element will get removed in shadow dom when conditionally rendering the elements and removing the source slot element in the light dom.

`experimentalSlotFixes` makes no difference **but using the "key" prop with a unique name fixes the issue.**

Consider the following basic example below.

### Expected Behavior

The Header Slot shall be removed in the example but the Footer Slot should be kept.

### System Info

```shell
npx stencil info

      System: node 20.10.0
    Platform: darwin (23.2.0)
   CPU Model: Apple M1 Pro (10 cpus)
    Compiler: /Users/michaelworm/projects/porsche-design-system/node_modules/@stencil/core/compiler/stencil.js
       Build: 1702322155
     Stencil: 4.8.2 🐳
  TypeScript: 5.2.2
      Rollup: 2.42.3
      Parse5: 7.1.2
      Sizzle: 2.42.3
      Terser: 5.26.0
```

### Steps to Reproduce

When clicking on "Remove Header" the Footer Slot will get removed and the Header Slot will be empty.
```html
<p-example>
  <div slot="header">
    <h1>Header Slot</h1>
  </div>

  <p>Default Slot for Content</p>
  <button onclick="document.querySelector('[slot=header]').remove()">Remove Header</button>

  <div slot="footer">
    <p>Footer Slot</p>
  </div>
</p-example>
```

```js
import { Component, Element, forceUpdate, h, type JSX } from '@stencil/core';

@Component({
  tag: 'p-example',
  shadow: true,
})
export class Example {
  @Element() public host!: HTMLElement;

  public render(): JSX.Element {
    const hasHeader = !!this.host.querySelector('[slot="header"]');
    const hasFooter = !!this.host.querySelector('[slot="footer"]');

    return (
      <div>
        {hasHeader && <slot name="header" onSlotchange={() => forceUpdate(this.host)} />}
        <slot />
        {hasFooter && <slot name="footer" onSlotchange={() => forceUpdate(this.host)} />}
      </div>
    );
  }
}
```
It seems that when adding elements between the slots, the bug does not occur.
```html
...
<div>
  {hasHeader && <slot name="header" onSlotchange={() => forceUpdate(this.host)} />}
  <div></div>
  <slot />
  <div></div>
  {hasFooter && <slot name="footer" onSlotchange={() => forceUpdate(this.host)} />}
</div>
...
```

### Code Reproduction URL

https://stackblitz.com/edit/node-dhtsl3?file=package.json,src%2Fcomponents%2Fmy-comp.tsx

### Additional Information

Remove the "key" props in the StackBlitz Demo to see the issue.

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
Putting up the code for this now so we can start looking at it!

## What is the current behavior?

We have a number of issues in the backlog relating to shortcomings in how children are reconciled across virtual DOM re-renders at present. For example, #1968 is caused by an interaction between a failure to correctly recognize the same child across a re-render and another issue with slot content relocation in a `scoped` context.

This implements an approach to addressing a subset of these issues by automatically inserting `key` attributes at compile-time, making it easier at run-time for the virtual DOM to determine when two children of a virtual DOM node should be matched up with each other.

## What is the new behavior?

Now as part of transpilation Stencil will insert randomly-generated `key` attributes into JSX syntax tree nodes. This allows the more-accurate key-based child reconciliation algorithm to be used almost all the time, fixing some edge cases where previously the Stencil runtime would accidentally misplace a child, failing to, for instance, correctly match up an old and new vdom node when nodes were reordered.

- fixes #1968 
- fixes #5263 

## Does this introduce a breaking change?

I don't believe it does!

- [ ] Yes
- [x] No

## Testing

I have added some tests which exercise the change and also ensured that all of our existing tests are passing without issue.

I also built and ran this in Framework, where all of the unit tests are passing and some smoke testing of components (the datetime in particular) doesn't reveal any potential problems.

But! For testing this I think it would be a good idea to check out the reproduction case from   #1968. Then:

- confirm you can reproduce the issue with `@stencil/core@latest`
- checkout this branch
- `npm run build && npm pack`
- install the tarball in the reproduction case
- confirm that the issue is fixed

### CODE REVIEW

File: src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { transpileModule } from '../test/transpile';
+import { formatCode } from '../test/utils';
+import { performAutomaticKeyInsertion } from '.';
+
+function transpile(code: string) {
+  return transpileModule(code, null, null, [performAutomaticKeyInsertion]);
+}
+
+describe('convert-decorators', () => {
```
Comment (alicewriteswrongs) [inline src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:9]: these aren't real tests right now, more just for debugging the impl

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    ...jsxElement.attributes.properties,
+  ]);
+
+  if (jsxElement.tagName.getText() === 'rect' || jsxElement.tagName.getText() === 'svg') {
+    // TODO diagnose the issue with svg tags
+    return jsxElement;
+  }
+
+  if (jsxElement.tagName.getText() === '') {
+    // TODO do we need this check?
```
Comment (alicewriteswrongs) [inline src/compiler/transformers/automatic-key-insertion/index.ts:149]: I don't know if this is necessary, leaving it in for now so I remember to check later

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  const attrs = ts.factory.createJsxAttributes([
+    ts.factory.createJsxAttribute(
+      ts.factory.createIdentifier('key'),
+      ts.factory.createStringLiteral(deriveJSXKey(jsxElement)),
+    ),
+    ...jsxElement.attributes.properties,
+  ]);
+
+  if (jsxElement.tagName.getText() === 'rect' || jsxElement.tagName.getText() === 'svg') {
+    // TODO diagnose the issue with svg tags
```
Comment (alicewriteswrongs) [inline src/compiler/transformers/automatic-key-insertion/index.ts:144]: svg attrs are _sometimes_ being dropped without this check, I'm not sure why but I'm going to leave this for now so I remember to investigate fully and then fix that issue

File: src/runtime/test/render-vdom.spec.tsx [hunk 4]
```diff
@@ -337,7 +337,7 @@ describe('render-vdom', () => {
         vdomXlink: false,
         vdomClass: false,
         vdomStyle: false,
-        vdomKey: false,
+        vdomKey: true,
```
Comment (alicewriteswrongs) [inline src/runtime/test/render-vdom.spec.tsx:340]: these tests for `vdomKey` just all changed because now every build has `key` attrs in it

File: test/karma/test-app/scoped-conditional/cmp.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        {/* This div below must be there too */}
+        <div>
+          before slot-&gt;
+          <slot />
+          &lt;-after slot
+        </div>
+      </div>
+    );
+  }
+}
```
Comment (alicewriteswrongs) [inline test/karma/test-app/scoped-conditional/cmp.tsx:31]: this is from a commit created by @rwaskiewicz which reproduces #1968 as a karma test.

Comment (alicewriteswrongs) [reply test/karma/test-app/scoped-conditional/cmp.tsx:31]: (ditto for the other stuff in this directory)

before adding the transformer these tests were failing, now they pass!

**Review by alicewriteswrongs — COMMENTED** [review]
noting a few things

**Comment by alicewriteswrongs** [comment]
I just published `@stencil/core@4.8.1-dev.1701985394.df7b575` from this branch, so that can be used to assess whether this addresses open issues

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

File: test/karma/test-app/slot-reorder/karma.spec.ts [hunk 6]
```diff
@@ -72,7 +75,7 @@ describe('slot-reorder', () => {
       expect(r.children[0].hasAttribute('hidden')).toBe(false);
       expect(r.children[0].getAttribute('name')).toBe('slot-b');
       expect(r.children[1].textContent.trim()).toBe('fallback default');
-      expect(r.children[1].hasAttribute('hidden')).toBe(false);
+      expect(r.children[1].hasAttribute('hidden')).toBe(true);
```
Comment (rwaskiewicz) [inline test/karma/test-app/slot-reorder/karma.spec.ts:78]: I'm not sure I fully understand this - the implication of this code change to me is that the `'hidden'` attribute's presence/value would be changing from today's existing behavior for scoped components - can you help me understand if this is the case / why we need to change this value?

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

**Comment by rwaskiewicz** [comment]
Ahh - two more things:
1. I think we should mark this in a minor release of Stencil/mark it as a feature. Can we update the PR name/commit messages?
2. Can we add some documentation around this behavior for the Stencil Site?

File: src/compiler/transformers/automatic-key-insertion/utils.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * @param jsxElement a typescript JSX syntax tree node which needs a key
+ * @returns a computed unique key for that element
+ */
+export function deriveJSXKey(jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement): string {
+  const hash = createHash('sha1')
+    .update(`${incrementer.next().value}__${jsxElement.tagName}__${jsxElement.pos}_${jsxElement.end}`)
+    .digest('hex')
+    .toLowerCase();
+  return hash;
+}
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/utils.ts:36]: I have some numbers on the JIRA ticket (STENCIL-893) and the tl;dr is that there's no discernible or significant performance difference when building Framework

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+   * method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findClassDeclVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isClassDeclaration(node)) {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findClassDeclVisitor, transformCtx);
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:49]: possibly not, just sort of a completeness / correctness thing I guess

but yeah since a class decl will always be a top-level syntax node with a `SourceFile` we can probably just return here, which will cut out some tree-traversal we'd do otherwise

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+   * This middle visitor function is responsible for finding the render method
+   * on a Stencil class and then passing off responsibility to the inner-most
+   * visitor, which deals with syntax nodes inside the method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findRenderMethodVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isMethodDeclaration(node) && node.name.getText() === 'render') {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:63]: mm that's a good point, we probably would want to do something like that anyway to avoid transforming someone's weird in-a-stencil-project React components or something

I'll pull those lines form `convert-decorators` out into a helper (`classDeclIsStencilComponent` or something)

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+function addKeyAttr(
+  jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement,
+): ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  const updatedAttributes = ts.factory.createJsxAttributes([
+    ts.factory.createJsxAttribute(
+      ts.factory.createIdentifier('key'),
+      ts.factory.createStringLiteral(deriveJSXKey(jsxElement)),
+    ),
+    ...jsxElement.attributes.properties,
+  ]);
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:119]: it won't because we splat in the existing attrs after the new one

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:119]: ah so this actually works more strangely that one might think!

I think TypeScript is perfectly fine with you doing

```tsx
<foo-bar key="first" key="second" />
```

because what it means to have multiple keys with the same name is really a runtime question

I'll add a check to ensure there's not already a property with the name `"key"` before we add

File: src/runtime/test/render-vdom.spec.tsx [hunk 11]
```diff
@@ -902,15 +902,17 @@ describe('render-vdom', () => {
...
       expect(rect.getAttribute('transform')).toBe(null);
 
       root.isOpen = true;
       await waitForChanges();
+      rect = root.querySelector('rect');
       expect(rect.getAttribute('transform')).toBe('rotate(45 27 27)');
 
       root.isOpen = false;
       await waitForChanges();
+      rect = root.querySelector('rect');
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/runtime/test/render-vdom.spec.tsx:915]: a small difference in how the elements get matched up means that where before a new `rect` element wouldn't be created on re-render it now will be, so the reference to the `rect` element created on the first render isn't to the updated element created on the second re-render

File: src/compiler/transformers/automatic-key-insertion/utils.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import ts from 'typescript';
+
+/**
+ * An incrementing-number generator, just as a little extra 'uniqueness'
+ * insurance for {@link deriveJSXKey}
+ */
+const incrementer = (function* () {
+  let val = 0;
+  while (true) {
+    yield val++;
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/utils.ts:12]: haha yeah that would be a big stencil project!

File: src/compiler/transformers/automatic-key-insertion/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:49]: I actually found out that we do support non-top-level Stencil components in at least one important case - our own tests files, like here for instance:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/runtime/test/render-vdom.spec.tsx#L8-L28

I think if we want to continue to support those we'll need to keep traversing the whole tree, unless we want to bite off some more added complexity

File: src/compiler/transformers/automatic-key-insertion/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:63]: So! I wasn't able to use the code from `convert-decorators` here because the relevant syntax nodes have already been removed from the component class node when they're 'static-ized' in `convertDecoratorsToStatic`, _however_ I was able to add a check which looks for the `is` property, as we already do here in `parseStaticComponentMeta`:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/compiler/transformers/static-to-meta/component.ts#L49-L53

this lets me determine whether or not a class is a Stencil component, and gate the transformation on that. I added a test to ensure that a non-decorated class which has a `render()` method with JSX doesn't get transformed.

File: src/compiler/transformers/automatic-key-insertion/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:119]: I added that check and also a test that if you've already set a `key` attr we won't add another one

File: src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    return h('div', { key: 'first-key' }, h('img', { key: 'second-key', src: 'my-img.png' }));
+  }
+  static get is() {
+    return 'cmp-a';
+  }
+}
+`,
+    );
+  });
+});
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:166]: good idea, I added a test to make sure that we don't add keys in cases where we can't assure ourselves it's safe to do so

File: test/karma/test-app/slot-reorder/karma.spec.ts (hunk 6 above)
Comment (alicewriteswrongs) [reply test/karma/test-app/slot-reorder/karma.spec.ts:78]: I believe what's going on is that previously this element was being destroyed and re-created so that it didn't have the hidden attribute anymore, now it is preserved across re-renders so although the _value_ of `el.getAttribute('hidden')` is I believe consistent and what it's supposed to be we see a difference in whether or not the attribute has been set on the element. Does that make sense?

File: src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+      this.todos.map((todo) => h('div', null, todo)),
+    );
+  }
+  static get is() {
+    return 'cmp-a';
+  }
+}
+`,
+    );
+  });
```
Comment (alicewriteswrongs) [inline src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:233]: this is a test to ensure that we don't add static keys to dynamic generated elements. I did a bit of a conservative cut-off for these - we don't attempt to transform any JSX nodes which are found within a JSX expression, so with (contrived) code like

```ts
@Component({ tag: 'cmp-a' })
class CmpA {
  render() {
    return <div>{ (() => <div>inner</div>)() }</div>
  }
}
```

we won't transform the 'inner' div in the IIFE because, since it will be generated at runtime, we could run into a lot of cases where we can't _really_ know too much about what's going on there and when it would not be safe to add keys.

**Review by alicewriteswrongs — COMMENTED** [review]
just noting a few things

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz I believe this is ready for another look! I've addressed your comments and added a few more tests to cover some more edge cases, lmk what you think!

File: test/karma/test-app/slot-reorder/karma.spec.ts (hunk 6 above)
Comment (rwaskiewicz) [reply test/karma/test-app/slot-reorder/karma.spec.ts:78]: I'm not sure, I'm still not totally clear as to how this is a result of automatic key insertion. Who/what was doing the recreation before, and who's doing it now? I suppose what I'm trying to understand here is any differences in our render cycle that might be a result of this change, and this question is just a small part of that.

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/automatic-key-insertion/index.ts [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+    case ts.SyntaxKind.JsxNamespacedName:
+      // this is a JSX attribute name like `foo:bar`
+      // see https://facebook.github.io/jsx/#prod-JSXNamespacedName
+      return attr.name.getText();
+    case ts.SyntaxKind.ComputedPropertyName:
+      const expression = attr.name.expression;
+      if (ts.isStringLiteral(expression) || ts.isNumericLiteral(expression)) {
+        return expression.text;
+      }
+    default:
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/compiler/transformers/automatic-key-insertion/index.ts:194]: yeah that's a good idea 👍

File: src/runtime/test/render-vdom.spec.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/runtime/test/render-vdom.spec.tsx:915]: It's possible if a stencil user is writing very detailed tests that assert that within a particular component a given HTML element is either destroyed and recreated _or_ that it's conserved across re-renders they might need to rewrite some tests. I have tested in Framework and all the tests there pass with no issues, so that's a good sign at least. Possibly we want to alert users to this? I'm not sure I would really want to classify it as a breaking change though since I think that we are fully within our rights to 'claim' the specific behavior of how Stencil's vdom produces the actual DOM that the user declaratively specifies as an implementation detail, and not itself part of the public API for building Stencil components and so on. I also haven't been able to produce counter-examples for myself which would show different user-facing behavior with this change, only 'under the hood' differences, and I think it's likely that most users wouldn't need to edit their tests with this change

File: test/karma/test-app/slot-reorder/karma.spec.ts (hunk 6 above)
Comment (alicewriteswrongs) [reply test/karma/test-app/slot-reorder/karma.spec.ts:78]: I realized that this is another situation where I think we want to bail before adding keys to the JSX. This is the component that's being tested in this file:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/test/karma/test-app/slot-reorder/cmp.tsx#L1-L40

In particular, note the `render()` method has two returns, each of which wrap some slot stuff with `<div>` tags, like so:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered">
          { ... }
        </div>
      );
    }

    return (
      <div>
        { ... }
      </div>
    );
  }
}
```

in this case we can't safely add `key` attrs to those divs, because doing so will cause the runtime to not match them up across renders. In particular, our implementation of `isSameVnode` assumes that _if keys are set_ on two vnodes then those are authoritative:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/src/runtime/vdom/vdom-render.ts#L586-L604

The transformed version of the component in question here would look something like this, with keys inserted:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered" key="asdfasdf1234">
          { ... }
        </div>
      );
    }

    return (
      <div key="hjklhjkl5678">
        { ... }
      </div>
    );
  }
}
```

because the keys set on those syntax nodes are _not_ equal, when the vdom assesses whether or not these nodes are equal it will see that they _have_ `key` attrs and, in that case, it will say "ok, are these equal? no? then these are not the same vnode". This will cause it to then re-render the whole thing, leading to some undesirable behavior where the `<div>` actually _should_ be kept around between re-renders so that its children can be diffed as well, rather than always being re-rendered on each render cycle.

I thought about this a bit, and I think that there's just not really a safe way that we can add keys in a case like this where there are multiple returns from the component's `render()` method. Or, at the least, I don't think we could safely do so without adding some more fancy static analysis that I dont _think_ we want to take on. Like adding keys would be safe here I think:

```tsx
export class ImageOrText {
  @Prop() showImage = false;

  render() {
    if (this.showImage) {
      return (
        <img src="..." />
      );
    }

    return (
      <span>text instead!</span>
    );
  }
}
```

because the two tags are different the vdom would already have to re-render the elements when `this.showImage` changes, so we wouldn't run into problems adding keys to the `img` and `span` tags in the JSX. But! I don't think we probably want to bite off trying to figure out when it's safe to do this. So instead I think we need to not add keys when:

1. there are two returns in a component's `render` method and
2. a component's `render` method returns a ternary

Does that all make sense? I've added some further constraints to when we transform JSX nodes and it makes this test in pass without the changes I have in this file right now.

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Overall looks good!

I have two asks before we can merge this:
1. I had one change I'd like to make to the code in its current state (comment made as a part of this review)
2. Did we ever push the PR up containing docs around this functionality?

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz I removed those changes, thanks for checking if they were still necessary!

**Comment by alicewriteswrongs** [comment]
also I opened a documentation here: https://github.com/ionic-team/stencil-site/pull/1330

**Review by rwaskiewicz — APPROVED** [review]

**Review by christian-bromann — APPROVED** [review]
Struggled a bit to reproduce the problem but with the help of @alicewriteswrongs reproducible example I was able that the issue is resolved with this patch 👍
